lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -0,0 +1,409 @@
+      auto target = instRecord->getTarget();
+      if (dutModules.insert(target->getModule()).second)
+        worklist.push_back(target);
+    }
+  }
+}
+
+/// Process an extraction annotation on an instance into a corresponding
+/// `ExtractionInfo` and a spot on the worklist for later moving things around.
+void ExtractInstancesPass::collectAnno(InstanceOp inst, Annotation anno) {
+  LLVM_DEBUG(llvm::dbgs() << "Processing instance `" << inst.name() << "` "
+                          << anno.getDict() << "\n");
+
+  auto getStringOrError = [&](StringRef member) {
+    auto attr = anno.getMember<StringAttr>(member);
+    if (!attr) {
+      inst.emitError("missing `")
+          << member << "` attribute in `" << anno.getClass() << "` annotation";
+      anyFailures = true;
+    }
+    return attr;
+  };
+
+  if (anno.isClass(extractBlackBoxAnnoClass)) {
+    auto filename = getStringOrError("filename");
+    auto prefix = getStringOrError("prefix");
+    auto dest = anno.getMember<StringAttr>("dest"); // optional
+    if (!filename || !prefix)
+      return;
+
+    // Find the parent module into which this instance should be propagate
+    // upwards. This does not include the optional wrapper module which gets
+    // created as a separate step.
+    // TODO: Is this even needed?
+
+    ExtractionInfo info;
+    info.traceFilename = filename;
+    info.prefix = prefix;
+    info.wrapperModule = (dest ? dest.getValue() : "");
+    // info.upwardsTarget = ;
+    extractionWorklist.push_back({inst, info});
+    return;
+  }
+}
+
+/// Move instances in the extraction worklist upwards in the hierarchy. This
+/// iteratively pushes instances up one level of hierarchy until they have
+/// arrived in the desired container module.
+void ExtractInstancesPass::extractInstances() {
+  // The list of ports to be added to an instance's parent module. Cleared and
+  // reused across instances.
+  SmallVector<std::pair<unsigned, PortInfo>> newPorts;
+  // The uniquified wiring prefix for each instance.
+  DenseMap<Operation *, SmallString<16>> instPrefices;
+  // The number of instances with the same prefix. Used to uniquify prefices.
+  DenseMap<StringRef, unsigned> prefixUniqueIDs;
+
+  while (!extractionWorklist.empty()) {
+    InstanceOp inst;
+    ExtractionInfo info;
+    std::tie(inst, info) = extractionWorklist.pop_back_val();
+    auto parent = inst->getParentOfType<FModuleOp>();
+    assert(parent && "instance expected to be in a module");
+
+    // If the instance is already in the right place (outside the DUT or already
+    // in the root module), there's nothing left for us to do. Otherwise we
+    // proceed to bubble it up one level in the hierarchy and add the resulting
+    // instances back to the worklist.
+    if (!dutModules.contains(parent) ||
+        instanceGraph->lookup(parent)->noUses()) {
+      LLVM_DEBUG(llvm::dbgs() << "\nNo need to further move " << inst << "\n");
+      continue;
+    }
```

Comment:
I think the logic here needs to put this into the DUT not move it outside the DUT? The interaction with `InjectDUTHierarchyAnnotation` moves all logic inside the DUT into a wrapper instance inside the DUT, but does not move the `MarkDUTAnnotation` to the wrapper. I think this logic will actually move things up one level too high?

Reply:
This part is tricky. In the absence of `InjectDUTHierarchyAnnotation`, this does the right thing. The `Simple` unit test in `ExtractBlackBoxesSpec` bubbles the black box all the way up into the test harness. That also seems to be the case for the other tests.

But then if `InjectDUTHierarchyAnnotation` is present, I see the corresponding unit test (`Multigrouping` of `ExtractClockGatesSpec`) actually bubbling the black boxes into the old "DUTModule", which then contains the "InjectedSubmodule". Things should work okay if the added DUT hierarchy moves the `InjectDUTHierarchyAnnotation` to the injected submodule, which would then cause extracted instances to sit next to the injected submodule _inside_ the original "DUTModule".

Does that make sense?

Reply:
Thanks for digging into the details. As long as it's doing the right thing when combined with `InjectDUTHierarchy`.

Reply:
Just found a few more details here: If the extraction annotation listed a wrapper module for the extracted instances (through the `dest` field), that wrapper was understood to be *within* the DUT itself. So the instances would either be extracted out of the DUT entirely, or they would be bubble up to the DUT root and grouoped in a wrapper module.